From 640b293046ed7eec41d5034a5c45c9c707d89347 Mon Sep 17 00:00:00 2001
From: Antoine Belvire <antoine.belvire@opensuse.org>
Date: Sat, 24 Aug 2019 14:14:07 +0200
Subject: [PATCH] Install translations for GUI

This adds a CMake install target for GUI qm files and allows the application to
load them from installation path.
---
 cmake/AddInstallTargets.cmake     | 1 +
 gui/CMakeLists.txt                | 1 +
 gui/cmake/AddDefines.cmake        | 3 +++
 gui/cmake/AddInstallTargets.cmake | 4 ++++
 gui/src/main.cpp                  | 6 +++++-
 5 files changed, 14 insertions(+), 1 deletion(-)
 create mode 100644 gui/cmake/AddDefines.cmake

diff --git a/cmake/AddInstallTargets.cmake b/cmake/AddInstallTargets.cmake
index 953497449..73710c4ae 100644
--- a/cmake/AddInstallTargets.cmake
+++ b/cmake/AddInstallTargets.cmake
@@ -15,6 +15,7 @@ else()
 endif()
 set(SOLARUS_MANUAL_INSTALL_DESTINATION "share/man" CACHE PATH "Manual install destination")
 set(SOLARUS_HEADERS_INSTALL_DESTINATION "include" CACHE PATH "Headers install destination")
+set(SOLARUSGUI_INSTALL_DATADIR "share/solarus-gui" CACHE PATH "GUI data install destination")
 
 # Files to install with make install.
 # Install the shared library and the solarus-run executable.
diff --git a/gui/CMakeLists.txt b/gui/CMakeLists.txt
index 7e3d80f3f..505ea1dfd 100644
--- a/gui/CMakeLists.txt
+++ b/gui/CMakeLists.txt
@@ -5,6 +5,7 @@ set(CMAKE_INCLUDE_CURRENT_DIR ON)
 # Tell CMake to run moc automatically when needed
 set(CMAKE_AUTOMOC ON)
 
+include(cmake/AddDefines.cmake)
 include(cmake/AddDependencies.cmake)
 include(cmake/AddIncludeDirectories.cmake)
 include(cmake/AddSolarusGuiLibrary.cmake)
diff --git a/gui/cmake/AddDefines.cmake b/gui/cmake/AddDefines.cmake
new file mode 100644
index 000000000..0d1aaa725
--- /dev/null
+++ b/gui/cmake/AddDefines.cmake
@@ -0,0 +1,3 @@
+# Add defines for the install path and the build path to help guess the assets
+# location at runtime
+add_definitions(-DSOLARUSGUI_TRANSLATION_PATH="${CMAKE_INSTALL_PREFIX}/${SOLARUSGUI_INSTALL_DATADIR}/translations")
diff --git a/gui/cmake/AddInstallTargets.cmake b/gui/cmake/AddInstallTargets.cmake
index d99f75a43..a229e6e36 100644
--- a/gui/cmake/AddInstallTargets.cmake
+++ b/gui/cmake/AddInstallTargets.cmake
@@ -12,6 +12,10 @@ install(FILES
   ${solarus-gui_FORMS_HEADERS}
   DESTINATION "${SOLARUS_HEADERS_INSTALL_DESTINATION}/solarus/gui"
 )
+install(FILES
+  ${solarus-gui_TRANSLATIONS_QM}
+  DESTINATION "${SOLARUSGUI_INSTALL_DATADIR}/translations"
+)
 
 # FreeDesktop compatible icons
 if(UNIX AND NOT APPLE)
diff --git a/gui/src/main.cpp b/gui/src/main.cpp
index 984c478f4..c8b602c74 100644
--- a/gui/src/main.cpp
+++ b/gui/src/main.cpp
@@ -73,7 +73,11 @@ int run_gui(int argc, char* argv[]) {
   application.installTranslator(&qt_translator);
 
   QTranslator translator;
-  translator.load("solarus_" + QLocale::system().name());
+  QString filename = "solarus_" + QLocale::system().name();
+  bool translation_loaded = translator.load(filename);
+  if (!translation_loaded) {
+    translator.load(filename, SOLARUSGUI_TRANSLATION_PATH);
+  }
   application.installTranslator(&translator);
 
   MainWindow window(nullptr);
-- 
2.22.0

